New Genome::Model::Tools::Analysis::Coverage::CoveragePlot added

gschang opened the pull request:
I have updated the whole source codes according to Chris's comments. Module dependency is solved by re-writing subroutines for external module replacement.

susannasiebert commented on lib/perl/Genome/Model/Tools/Analysis/Coverage/CoveragePlot.pm:
```diff
@@ -0,0 +1,1439 @@
+	    is => 'String',
+	    is_optional => 1,
+	    doc => 'Comma-separated list of bam files to generate coverage plots from, Must specify either --bam_files or --refalign_models',
+	},
```
If you specify this (and other comma-separated properties) as `is_many => 1` you don't have to do any splitting up into an array during execution. UR will do that for you when you call `$self->bam_files`.

susannasiebert commented on lib/perl/Genome/Model/Tools/Analysis/Coverage/CoveragePlot.pm:
```diff
@@ -0,0 +1,1439 @@
+	    is => 'Number',
+	    is_optional => 1,
+	    doc => 'Comma-separated list of refalign model IDs to generate coverage plots from, bam_files will override refalign_models if given',
+	},
```
You should make this a `is => 'Genome::Model::ReferenceAlignment'`. This will automatically get the model object for each id for you and it will do some error checking that the id is indeed a reference alignment model. You don't need to do the retrieval and check on lines 188 and 189. 

This should also be a `is_many => 1`

gschang replied:
Actually, this option does not work. It showed "ERROR: Unknown option: refalign-models". I also commented about this error message in the source code.

brummett commented on lib/perl/Genome/Model/Tools/Analysis/Coverage/CoveragePlot.pm:
```diff
@@ -0,0 +1,1439 @@
+    
+    # as default
+    die "roi-file-path required" unless -e $roi_file_path;
+    print STDERR sprintf("%d region(s) of interest from %s\n", `wc -l $roi_file_path | cut -d " " -f 1`, $roi_file_path);
```
All these prints to STDERR should instead use the methods ```debug_message()```, ```status_message()```, ```warning_message()``` or ```error_message()``` depending on the severity of the message.  For example, this one could be:
```
$self->status_message("%d region(s) of interest from %s\n",
                      `wc -l $roi_file_path | cut -d " " -f 1`,
                      $roi_file_path);
```

gschang replied:
As you suggested in the comment of "create_hash()" below, apipe does not want to create a class namespace arbitrarily and does not allow individual developer to push their own class into Genome Model library. So, I extracted from my class library and pasted only necessary subroutines in this tool. You are right, but I took it as a practical solution that is not for object-oriented programming here.

brummett replied:
```debug_message()``` and the others are methods of the base class, UR::Object.  They take the same arguments as ```printf()``` and are the standard way to communicate short messages to the user.

```debug_message``` is intended to be information to the developer.  Their messages are normally suppressed.
```status_message``` is what we normally use to print general information to the user.
```warning_message``` is for when something unusual or unexpected has happened.
```error_message``` is for when something dire has happened.  The program will usually die or exit after sending the message.

brummett replied:
create_hash() and its related methods should definitely be in another class.  Returning entities blessed into this Command module's class will be confusing to the next person to work on this, and is bound to cause trouble in the future.

If you don't want to create another .pm file for fear of cluttering up the filesystem, you could just declare another ```package``` at the bottom of this same file, and move the relevant methods down there.  Let us know if you need help making the changes.

brummett commented:
The ```create_hash()``` is actually a constructor for some new class and really belongs in its own package.  As it is, it is returning objects blessed into the ```Genome::Model::Tools::Analysis::Coverage::CoveragePlot``` package when they don't belong there.

Here are a couple of links that describes how to make a simple hash-based class in Perl:
http://stackoverflow.com/questions/1037312/how-do-i-create-a-perl-class
http://www.zentut.com/perl-tutorial/perl-oop/
All the methods that operate on those objects (like ```write()``` and ```from_file()```) need to be moved into the class' package as well.  It's also good practice to put each kind of thing in its own class.  For example, it looks like things you'd call ```from_file()``` on are a different kind of thing that those you'd call ```gradient_2colors()``` on

dufeiyu commented on lib/perl/Genome/Model/Tools/Analysis/Coverage/CoveragePlot.pm:
```diff
@@ -0,0 +1,1439 @@
+    my $wingspan = $self->wingspan;
+    
+    my $bw = $self->bw;
+    $bw = "nrd0" if ($bw =~ /^\s*$/);
```
This "nrd0" should be set as default_value when making "bw" parameter (line 64). Also to me "band_width" seems to be a better parameter name than "bw".

gschang commented:
Thank you for all your nice comments. I will address and get back to you again next week.
--Gue Su

chrisamiller commented on lib/perl/Genome/Model/Tools/Analysis/Coverage/CoveragePlot.R:
```diff
@@ -0,0 +1,1549 @@
+
+
+
+cs_range_limit3 <- function(x, y, z, xlim=NULL, ylim=NULL, zlim=NULL)
```
Many of these function names could be made more descriptive. The difference between cs_range_limit2 and cs_range_limit3 is not at all apparent from the function name.

chrisamiller commented on lib/perl/Genome/Model/Tools/Analysis/Coverage/CoveragePlot.R:
```diff
@@ -0,0 +1,1549 @@
+        #print(xmar);
+        #print(vp$x);
+        #print(ymar);
+        #print(vp$y);
```
Lots of comments that could be removed. Debugging statements should be taken out. If the above chunk that's labeled "The following margins do not work in an accurate way" is old code that will not be used, just remove it.

gschang replied:
Those comments were already cleaned up. I also realized some functions 
are unnecessary (never used) in the CoveragePlot code. So, I had updated 
the codes accordingly, but I don't know how to push them up onto the 
previous commitment (which you are now seeing). Let me know how to do by 
commands.


On 3/4/2015 11:57 AM, Chris Miller wrote:
>
> In lib/perl/Genome/Model/Tools/Analysis/Coverage/CoveragePlot.R 
> <https://github.com/genome/genome/pull/525#discussion_r25796352>:
>
> > +    }
> > +    else
> > +    {
> > +        # uses the given viewport
> > +        cv <- vp;
> > +
> > +        # the margins obtained from the given viewport
> > +        # The following margins do not work in an accurate way
> > +        #xmar = vp$x;
> > +        #ymar = vp$y;
> > +
> > +        # for debug
> > +        #print(xmar);
> > +        #print(vp$x);
> > +        #print(ymar);
> > +        #print(vp$y);
>
> Lots of comments that could be removed. Debugging statements should be 
> taken out. If the above chunk that's labeled "The following margins do 
> not work in an accurate way" is old code that will not be used, just 
> remove it.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/genome/genome/pull/525/files#r25796352>.
>



____
This email message is a private communication. The information transmitted, including attachments, is intended only for the person or entity to which it is addressed and may contain confidential, privileged, and/or proprietary material. Any review, duplication, retransmission, distribution, or other use of, or taking of any action in reliance upon, this information by persons or entities other than the intended recipient is unauthorized by the sender and is prohibited. If you have received this message in error, please contact the sender immediately by return email and delete the original message from all computer systems. Thank you.

gschang replied:
I have cleaned up the source codes. In fact, many unnecessary functions were pasted from my R grid library, which also makes a confusion in the function name. Now, it looks clearer.

chrisamiller commented on lib/perl/Genome/Model/Tools/Analysis/Coverage/CoveragePlot.R:
```diff
@@ -0,0 +1,1549 @@
+        
+        # creates a gridline
+        if (! is.null(lty))
+        {
```
In general, the preferred TGI style is to do:

    if(statement){
        code
    }

with the open parentheses on the same line as the statement (makes for more compact code).  I'm not going to ask you to clean all of these up at this point, but it's something to note next time you write a script.

chrisamiller commented on lib/perl/Genome/Model/Tools/Analysis/Coverage/CoveragePlot.R:
```diff
@@ -0,0 +1,1549 @@
+
+# --------------------------------------
+# gets the command line arguments when run by Rscript
+args <- commandArgs(trailingOnly=TRUE);
```
Wouldn't be a bad idea to verify that you get the expected number of arguments here.

gschang commented:
Hi, all. I have updated the source code of CoveragePlot with your comments. Some didn't work and I am going to add individual messages here. Looking forward to your next comments. Thank you for your comments again.

brummett commented on lib/perl/Genome/Model/Tools/Analysis/Coverage/CoveragePlot.pm:
```diff
@@ -0,0 +1,1375 @@
+}
+
+
+sub is_numeric {
```
The function ```looks_like_number``` in the module Scalar::Util does this same job and is more thorough.

ebelter replied:
@brummett  It that preferred to Regexp::Common?

brummett replied:
I guess in this case, you could use either.  Scalar::Util::looks_like_number() uses the same low-level, perl internal mechanism used when it converts a string to a number.  Any string that passes looks_like_number() won't generate warnings or exceptions if you try to use it in a numerical expression.

I'd try looks_like_number() first, and only use something more complicated like a regex if it didn't work.  A regex will always be slower.

gschang commented:
I appreciate your helpful comments. I addressed most comments and updated my source codes. Regarding your create_hash() comment, I solved it both by creating a new class (in fact, this class is a part of my original class CSLib::Tabular, and I pasted only necessary methods) and by converting some to normal subroutines, as you suggested. I want to see how you think.

brummett commented:
This looks much improved and removes my concerns.
+1

chrisamiller commented:
+1 let's merge this
